File: src/com/leozhi/topic852/Solution03.java
package com.leozhi.topic852;

/**
 * @author leozhi
 * 二分查找
 * 通过
 * 0ms
 */
public class Solution03 {
    public int peakIndexInMountainArray(int[] arr) {
        int left = 0, right = arr.length - 1;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (arr[mid] < arr[mid + 1]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }
}

File: src/com/leozhi/topic605/Solution03.kt
package com.leozhi.topic605

/**
 * @author leozhi
 * 贪心算法
 * 通过
 * 244ms
 */
class Solution03 {
    fun canPlaceFlowers(flowerbed: IntArray, n: Int): Boolean {
        if (n == 0) {
            return true
        }
        val len = flowerbed.size
        var count = 0
        for (i in 0 until len) {
            val left = i == 0 || flowerbed[i - 1] == 0
            val right = i == len - 1 || flowerbed[i + 1] == 0
            if (flowerbed[i] == 0 && left && right) {
                flowerbed[i] = 1
                count++
                if (count == n) {
                    return true
                }
            }
        }
        return false
    }
}